rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md

```diff
@@ -0,0 +1,115 @@
+# RFC 9460 - 2021-12-01 - Healthcheck endpoint improvements
+
+Our existing health endpoint is too limited and additional features are required
+to properly interact with load balancers or Kubernetes service discovery.
+
+## Context
+
+- [Additional response for api health endpoint](https://github.com/vectordotdev/vector/issues/9160)
+
+## Scope
+
+### In scope
+
+- Update health endpoint to improve Vector's reliability in production deployments
+- Update default configurations to improve Vector's out-of-the-box reliability
+
+### Out of scope
+
+- Moving the health endpoint out of our existing API
+- Adding user facing configuration to adjust endpoint behavior or responses
+
+## Pain
+
+- Today's health endpoint causes 502's when running behind a loadbalancer
+- The healthcheck isn't integrated with any components and only represents if Vector
+itself is running
+
+## Proposal
+
+### User Experience
+
+The existing health endpoint on Vector's API is enhanced to return 503's
+when Vector is shutting down. This additional response will allow load balancers
+and service discovery to better integrate with Vector, sources already begin
+rejecting requests received during the shutdown process and updating the health
+check will allow for removing the instance from available backends and avoid
+routing traffic to an instance that will end up rejecting the request regardless.
+
+### Implementation
+
+We have existing logic in place for other components to handle Vector's shutdown.
+The [health handler](https://github.com/vectordotdev/vector/blob/master/src/api/handler.rs#L7)
+can be updated to respond 503 if shutdown has started and 200 otherwise. This
+should be a minor change and not negatively impact existing deployments.
+
+## Rationale
+
+This improvement is low effort but greatly improves the experience of running
+Vector behind a load balancer or in a service discovery system. As we reccomend
+both of those options for production deployments there's very little reason not
+to implement this.
+
+## Drawbacks
+
+This is a minor change overall and does not add any real engineering burden.
```

Comment:
This may be out of scope for the initial RFC, but it's worth noting there's still _some_ engineering work.

Potentially:

- Moving API init above config validation (if we want 503s on startup). This is a little more involved than shifting a few lines further up the page; config validation occurs within `Application::prepare_from_opts` and the API is separately spawned in `Application::run`, with a dependency on config already being available. This would need adjusting if we want to issue 503s pre-readiness.

- Additional signal handling to determine when Vector is starting up or shutting down. Even if this winds up just being shutdown, we need some way of updating the health state so that it responds with 503. Currently this is a idempotent/immutable route.

Reply:
Yeah, certainly some "hand waving" on difficulty and maintainability here. It's definitely less involved than per component "health" and calculating some health score out of that 😄